### PR TITLE
added flow status for pod failure

### PR DESCRIPTION
**Pull request by hqin2016**
A new state will help identify flow failures due to pod (k8s) related failures.
Needs to account for state showing up correctly on UI
Tested on local dev box solo server and h6 web server.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/metrics/ContainerizationMetricsImpl.java [hunk 1]
```diff
@@ -53,6 +51,9 @@ public void setUp() {
     this.podReady = this.metricsManager.addMeter("Pod-Ready-Meter");
     this.podInitFailure = this.metricsManager.addMeter("Pod-Init-Failure-Meter");
     this.podAppFailure = this.metricsManager.addMeter("Pod-App-Failure-Meter");
+    this.flowSubmitToContainer = this.metricsManager.addMeter("Flow-Submit-to-Container-Meter");
```
Comment (aditya1105) [inline azkaban-common/src/main/java/azkaban/metrics/ContainerizationMetricsImpl.java:54]: Where are these meters incremented or utilized?

Comment (hqin2016) [reply azkaban-common/src/main/java/azkaban/metrics/ContainerizationMetricsImpl.java:54]: They should have been added in https://github.com/azkaban/azkaban/pull/2882.

File: azkaban-common/src/main/java/azkaban/executor/Status.java [hunk 2]
```diff
@@ -82,6 +85,7 @@ public static boolean isStatusRunning(final Status status) {
 
   public static boolean isStatusFailed(final Status status) {
     switch (status) {
+      case POD_FAILED:
```
Comment (svnarumugam) [inline azkaban-common/src/main/java/azkaban/executor/Status.java:88]: In conditional flows, this method is used to check if we need to run other nodes. when we combine the node failures with system-related failure (in this case pod), this will likely affect that feature. It's worth double-checking that conditional flow use case.

**Review by aditya1105 — APPROVED** [review]

**Review by svnarumugam — DISMISSED** [review]
LGTM. Thank a lot for addressing the comments, Hao!

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/WebUtils.java [hunk 3]
```diff
@@ -93,6 +93,8 @@ public static String formatStatus(final Status status) {
         return "Killing";
       case DISPATCHING:
         return "Dispatching";
+      case POD_FAILED:
```
Comment (jakhani) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/WebUtils.java:96]: Shall we rename this status to CONTAINER_FAILED? 
^^ cc: @sshardool

File: azkaban-web-server/src/main/less/flow.less [hunk 4]
```diff
@@ -364,6 +376,10 @@ li.tree-list-item {
       background-position: 48px 0px;
     }
 
+    &.POD_FAILED .icon {
```
Comment (jakhani) [inline azkaban-web-server/src/main/less/flow.less:379]: Can you please add new status in historypage.vm as well?

File: azkaban-web-server/src/web/js/azkaban/util/job-status.js [hunk 5]
```diff
@@ -16,7 +16,7 @@
 
 var statusList = ["FAILED", "FAILED_FINISHING", "SUCCEEDED", "RUNNING",
   "WAITING", "KILLED", "DISABLED", "READY", "CANCELLED", "UNKNOWN", "PAUSED",
-  "SKIPPED", "QUEUED", "FAILED_SUCCEEDED", "DISPATCHING"];
+  "SKIPPED", "QUEUED", "FAILED_SUCCEEDED", "DISPATCHING", "POD_FAILED"];
```
Comment (jakhani) [inline azkaban-web-server/src/web/js/azkaban/util/job-status.js:19]: Can you please validate that Kill link is not coming on Flow execution page and prepare execution link is coming instead? The behavior should be same as any other final state.

File: azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java [hunk 6]
```diff
@@ -249,7 +249,9 @@ public static void failEverything(final ExecutableFlow exFlow) {
       exFlow.setEndTime(time);
     }
 
-    exFlow.setStatus(Status.FAILED);
+    if (exFlow.getStatus() != Status.POD_FAILED) {
```
Comment (sshardool) [inline azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java:252]: Perhaps make this check more generic and switch the status to FAILED if the previous status is already a final one ?
Something like `if (!isStatusFinised(exFlow.getStatus())`

File: azkaban-common/src/main/java/azkaban/executor/Status.java [hunk 7]
```diff
@@ -30,6 +30,8 @@
   SUCCEEDED(50),
   KILLING(55),
   KILLED(60),
+  // POD_FAILED refers to a failed containerized flow due to pod failure
+  POD_FAILED(65),
```
Comment (sshardool) [inline azkaban-common/src/main/java/azkaban/executor/Status.java:34]: We should consider giving this state a more generic name which could be used for flows in similar state in non-containerized Azkban. For instance in the event of a bare-metal executor crashing we could the switch any flows assigned to that executor to this state. 
Something like `EXECUTE_INFRA_FAILED`, better alternatives should be possible though.

File: azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java [hunk 8]
```diff
@@ -122,12 +123,11 @@ private void validateAndPreProcess(AzPodStatusMetadata event) {
...
     logEventCacheStats(eventCount);
 
-    AzPodStatus currentStatus = getCurrentAzPodStatus(event.getPodName());
-    boolean isTransitionValid = validateTransition(event, currentStatus);
+    boolean isTransitionValid = validateTransition(event);
     if (!isTransitionValid) {
       IllegalStateException transitionException = new IllegalStateException(
           format("Pod status transition is not supported %s -> %s, for pod %s",
-              currentStatus,
+              getCurrentAzPodStatus(event.getPodName()),
```
Comment (sshardool) [inline azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java:130]: Why is this change needed ?

File: azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java [hunk 9]
```diff
@@ -239,6 +239,9 @@ private void updatePodStatus(AzPodStatusMetadata event) {
           "Flow execution-id %d for pod %s does not have a final status in database and will be "
               + "finalized.", executionId, event.getPodName()));
       final String reason = "Flow Pod execution was completed.";
+      if (!validateTransition(event)) {
```
Comment (sshardool) [inline azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java:242]: The method `compareAndFinalizeFlowStatus` should not be invoking `validateTransition` directly, would be better to let this method be narrow in scope which and thereby retain its reusability.  Also note that the check for `validateTransition` has already occurred before this method will be invoked and calling it again in this method could be a little confusing.

I recommend,

1. Add an additional argument `Status finalStatus` to all of these methods - `compareAndFinalizeFlow`, `finalizeFlowAndDeleteContainer`,  `ExecutionControllerUtils.finalizeFlow`. The default value for this argument can be `FAILED` if needed (using overloaded methods with this argument missing)
2. In the method `validateAndPreProcess`, once we have found that the transition is invalid then invoke `finalizeFlowAndDeleteContainer ` with the `finalStatus` argument set to `POD_FAILED`

File: azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java [hunk 10]
```diff
@@ -239,6 +239,9 @@ private void updatePodStatus(AzPodStatusMetadata event) {
...
+        executableFlow.setStatus(Status.POD_FAILED);
```
Comment (sshardool) [inline azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java:243]: Setting the flow status directly will bypass a number of additional tasks performed by `ExecutionControllerUtils.finalizeFlow`, including alerting users about flow failures. 
As mentioned in the previous comment, please see if the `finalizeFlow` can be updated to accept and additional param for `finalStatus`. We can discuss if that seem problematic for any reason.

**Review by sshardool — COMMENTED** [review]
Thanks a lot for making this change, this will be very useful.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/WebUtils.java (hunk 3 above)
Comment (sshardool) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/WebUtils.java:96]: `CONTAINER_FAILED` certainly seems better than `POD_FAILED` but I was hoping for name that could be use to reflect similar failures for bare-metal executors, e.g when the executor-server crashes. 
Any suggestions @jakhani ?  (or should we not try that at all)

File: azkaban-common/src/main/java/azkaban/executor/Status.java (hunk 7 above)
Comment (ypadron-in) [reply azkaban-common/src/main/java/azkaban/executor/Status.java:34]: +1 on using a generic name not indicative or coupled to a specific implementation.

**Review by ypadron-in — CHANGES_REQUESTED** [review]
@hqin2016 Could you please enumerate some example scenarios where an execution will finish with the new status?

I may be missing some context here but I feel we're mixing 2 things with the new status: `execution status` and `reason for failure`.

Reason for failure can be anything from network failures to Azkaban server failures to user code failures, we can't possibly list them all.

Why do we think we should create a status that indicates an infrastructure failure? 
What will prevent us in the future to keep adding failed status with associated reason for failure? 
Wouldn't execution status become unreliable and a maintenance nightmare for users who depend on them to do X or Y? For example: today users only need to monitor FAILED status to determine if an execution failed (for whatever reason). With this new status users could miss failed executions if they don't update their code to start monitoring the new status as well.

**Review by hqin2016 — COMMENTED** [review]
> @hqin2016 Could you please enumerate some example scenarios where an execution will finish with the new status?
> 
> I may be missing some context here but I feel we're mixing 2 things with the new status: `execution status` and `reason for failure`.
> 
> Reason for failure can be anything from network failures to Azkaban server failures to user code failures, we can't possibly list them all.
> 
> Why do we think we should create a status that indicates an infrastructure failure?
> What will prevent us in the future to keep adding failed status with associated reason for failure?
> Wouldn't execution status become unreliable and a maintenance nightmare for users who depend on them to do X or Y? For example: today users only need to monitor FAILED status to determine if an execution failed (for whatever reason). With this new status users could miss failed executions if they don't update their code to start monitoring the new status as well.

The purpose of creating such PR and status is described in BDP-6898. If you have time, would you like to join the next focus group meeting?

File: azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java [hunk 11]
```diff
@@ -238,7 +240,10 @@ private void updatePodStatus(AzPodStatusMetadata event) {
       logger.info(format(
           "Flow execution-id %d for pod %s does not have a final status in database and will be "
               + "finalized.", executionId, event.getPodName()));
-      final String reason = "Flow Pod execution was completed.";
+      final String reason = "Flow pod execution was completed.";
+      if (finalStatus == Status.EXECUTION_STOPPED) {
```
Comment (sshardool) [inline azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java:244]: Looks like the argument `finalStatus` will be meaningful only when its value is `EXECUTION_STOPPED` and will be completely ignored otherwise. 
How about one of the following alternatives:
(1) Don't ignore the argument and if the flow is being finalized then always use the given `finalStatus` 
(2) Change the argument type to `Optional<Status> finalStatus` and use the argument only when `finalStatus.isPresent() == true`

**Review by sshardool — DISMISSED** [review]
Thanks for incorporating the comments!

File: azkaban-common/src/main/java/azkaban/executor/Status.java [hunk 12]
```diff
@@ -30,6 +30,8 @@
   SUCCEEDED(50),
   KILLING(55),
   KILLED(60),
+  // EXECUTION_STOPPED refers to a terminating flow status due to crashed executor/container
```
Comment (djaiswal83) [inline azkaban-common/src/main/java/azkaban/executor/Status.java:33]: terminal flow status?

File: azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java [hunk 13]
```diff
@@ -283,11 +288,12 @@ private void deleteFlowContainer(AzPodStatusMetadata event) {
 
   /**
    * Finalize the flow for the given event and delete its container.
-   *
-   * @param event
+   *  @param event
```
Comment (djaiswal83) [inline azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java:291]: nit: indentation

**Review by djaiswal83 — CHANGES_REQUESTED** [review]
I have couple of minor comments. Feel free to ignore them if time sensitive.

**Review by djaiswal83 — DISMISSED** [review]
Thanks for addressing the comments.

File: azkaban-common/src/main/java/azkaban/executor/Status.java [hunk 14]
```diff
@@ -30,6 +30,8 @@
   SUCCEEDED(50),
   KILLING(55),
   KILLED(60),
+  // EXECUTION_STOPPED refers to a terminal flow status due to crashed executor/container
+  EXECUTION_STOPPED(65),
```
Comment (ypadron-in) [inline azkaban-common/src/main/java/azkaban/executor/Status.java:34]: Could the name of the status be confusing to users? It says the execution stopped but is it a temporary stop? Will it continue to run at some point? To me the name does not suggest a final status but a transition one. WDYT @sshardool @jakhani @djaiswal83 @aditya1105

File: azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java [hunk 15]
```diff
@@ -132,7 +132,7 @@ private void validateAndPreProcess(AzPodStatusMetadata event) {
               event.getPodName()));
       logger.error("Unsupported state transition.", transitionException);
       try {
-        finalizeFlowAndDeleteContainer(event);
+        finalizeFlowAndDeleteContainer(event, Optional.of(Status.EXECUTION_STOPPED));
```
Comment (ypadron-in) [inline azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java:135]: In the case of Pod going down due to OOM caused by user code. Will the execution be marked with status FAILED or EXECUTION_STOPPED? @sshardool @jakhani @djaiswal83 @aditya1105

File: azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java [hunk 16]
```diff
@@ -283,11 +288,12 @@ private void deleteFlowContainer(AzPodStatusMetadata event) {
 
   /**
    * Finalize the flow for the given event and delete its container.
-   *
    * @param event
+   * @param finalStatus
    */
-  private void finalizeFlowAndDeleteContainer(AzPodStatusMetadata event) {
-    Optional<Status> originalFlowStatus = compareAndFinalizeFlowStatus(event);
+  private void finalizeFlowAndDeleteContainer(AzPodStatusMetadata event,
```
Comment (ypadron-in) [inline azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusManagerListener.java:294]: This PR implements the new status for containerized executions only but bare metal executors also crash so we should be consistent and implement it for bare metal executors too. Could you please make the necessary changes?
So far execution failures due to executor crashes has been marked as FAILED. When this PR is merged their final status should change to EXECUTION_STOPPED which affects ALL Azkaban users and potentially break their code. Users will need to modify their code and logic to account for the new status. Are we ok with that? @sshardool @jakhani @djaiswal83 @aditya1105

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/WebUtils.java [hunk 17]
```diff
@@ -93,6 +93,8 @@ public static String formatStatus(final Status status) {
         return "Killing";
       case DISPATCHING:
         return "Dispatching";
+      case EXECUTION_STOPPED:
+        return "Execution stopped, crashed executor/container";
```
Comment (ypadron-in) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/WebUtils.java:97]: How is this message used? Can we remove the "crashed executor/container" part which is an implementation detail?
